File: backend_node/api/files.js
const express = require('express');
const path = require('path');
const fs = require('fs');

const router = express.Router();

const PROJECT_ROOT = path.join(__dirname, '..', '..');

router.get('/view_file/*', (req, res) => {
    try {
        // Get the relative path from the URL
        const relPath = req.params[0];
        const absPath = path.resolve(path.join(PROJECT_ROOT, relPath));
        
        // Security: Only allow files inside project directory
        if (!absPath.startsWith(PROJECT_ROOT)) {
            return res.status(403).json({ error: 'Access denied' });
        }
        
        if (!fs.existsSync(absPath)) {
            return res.status(404).json({ error: 'File not found' });
        }
        
        // Send the file
        res.sendFile(absPath);
        
    } catch (error) {
        console.error('Error serving file:', error);
        res.status(500).json({ error: error.message });
    }
});

module.exports = router;

File: backend/api/files.py
from flask import Blueprint, send_file
import os

PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))

files_bp = Blueprint('files', __name__)

@files_bp.route('/view_file/<path:rel_path>')
def view_file(rel_path):
    """Serve project files for viewing"""
    try:
        abs_path = os.path.abspath(os.path.join(PROJECT_ROOT, rel_path))
        
        # Security: Only allow files inside project directory
        if not abs_path.startswith(PROJECT_ROOT):
            return {'error': 'Access denied'}, 403
        
        if not os.path.exists(abs_path):
            return {'error': 'File not found'}, 404
            
        return send_file(abs_path)
    except Exception as e:
        return {'error': str(e)}, 500
